### PR TITLE
Add Light Scripting API and deprecate the AdditionalShadowData component

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
+ Add the light scripting API (see link below)
+ Deprecated the `AdditionalShadowData` component and moved all its properties to `HDAdditionalLightData`

You can have an overview of the scripting API here: [HDAdditionalLightData_Doc.pdf](https://github.com/Unity-Technologies/ScriptableRenderPipeline/files/3351728/screencapture-file-D-LWRP-Documentation-html-class-unity-engine-1-1-experimental-1-1-rendering-1-1-h-d-pipeline-1-1-h-d-additional-light-data-html-2019-07-02-20_06_40.pdf)

Example code:
```csharp
            var go = new GameObject("HD Light");
            var hdLight = go.AddHDLight(HDLightType.BoxSpot);

            // Set light parameters
            hdLight.SetIntensity(05);
            hdLight.SetRange(1.01f);
            hdLight.SetSpotAngle(60);
            hdLight.SetCookie(cookie2D);

            // Enable shadows
            hdLight.EnableShadows(true);
            hdLight.SetShadowResolution(Mathf.NextPowerOfTwo(Random.Range(32, 512)));
```

---
### Testing status
**Katana Tests**: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2Ffeature%2Flight_scripting&automation-tools_branch=master&unity_branch=trunk

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**:  Added a test that spawns lights in different configurations:
Set Color, Set Intensity using units, Set Cookie, Set Range, Set Light Unit, Set Color Temperature, Set Outer Angle | Inner Angle, Set area light size, Set box spot light size, Set Volumetric Dimmer, Set Shadow resolution, Set Contact Shadows, Set Light Layer, Toggle affect diffuse, Toggle Range Attenuation.

On every kind of light type (Directional, Point, Box Spot, Pyramid Spot, Cone Spot, Rectange and Tube).
image

---
### Overall Product Risks
**Technical Risk**: Medium [Rémi: NOOOO Way Higher]

**Halo Effect**: High

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+namespace UnityEngine.Experimental.Rendering.HDPipeline
+{
+    public partial class HDAdditionalLightData : ISerializationCallbackReceiver
+    {
+        // TODO: Use proper migration toolkit
+        // 3. Added ShadowNearPlane to HDRP additional light data, we don't use Light.shadowNearPlane anymore
+        // 4. Migrate HDAdditionalLightData.lightLayer to Light.renderingLayerMask
+        // 5. Added the ShadowLayer
+        private const int currentVersion = 5;
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: Please use enum for versioning. It is simple to read and also to understand what is behind a version number.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        [HideInInspector, SerializeField]
+        [FormerlySerializedAs("m_Version")]
+        [System.Obsolete("version is deprecated, use m_Version instead")]
+        private float version = currentVersion;
+        [SerializeField]
+        private int m_Version = currentVersion;
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: This serialization pattern is odd.
version will always be filled with m_Version value.
If version is not used anymore, as it is private, removing it.

Comment (RSlysz) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: I am not sure how the conflict is solved along the type change but same serialization name.

Comment (RSlysz) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: If we can drop the really first float versions (EXPERIMENTAL), please use the migration toolkit for better consistency along HDRP code

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.cs [hunk 3]
```diff
@@ -30,6 +33,36 @@ public enum LightUnit
...
+        Lux = LightUnit.Lux,
+        Ev100 = LightUnit.Ev100
+    }
+
+    /// <summary>
+    /// Shadow Update mode
+    /// </summary>
+    public enum ShadowUpdateMode
+    {
+        EveryFrame = 0,
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.cs]: unneeded. First item of enum is always 0 in c#

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Light/HDAdditionalLightDataEditorExtension.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+using UnityEditor.Experimental.Rendering.HDPipeline;
+using UnityEngine.Serialization;
+using UnityEngine.Experimental.Rendering.HDPipeline;
+
+namespace UnityEditor.Experimental.Rendering.HDPipeline
+{
+    // Editor only functions for HDAdditoonalLightData User API
+    public static class HDAdditionalLightDataEditorExtension
+    {
+        public static void EnableColorTemperature(this HDAdditionalLightData hdLight, bool enable)
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/Light/HDAdditionalLightDataEditorExtension.cs]: missing inline documentation for public API

File: com.unity.render-pipelines.high-definition/Editor/Lighting/AdditionalShadowDataEditor.cs [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 namespace UnityEditor.Experimental.Rendering.HDPipeline
 {
+#pragma warning disable 618 // Obsolete warning
```
Comment (RSlysz) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/AdditionalShadowDataEditor.cs]: If it is not migrated with removal, we do not need to keep the editor here.
Else considere adding migration step to clean scene with uneeded shadow data

File: TestProjects/HDRP_Tests/Assets/GraphicTests/Scenes/9x_Other/9500_LightScripting/trasnparnetShadowCasting.mat [hunk 6]
```diff
@@ -0,0 +1,242 @@
+%YAML 1.1
```
Comment (sebastienlagarde) [inline TestProjects/HDRP_Tests/Assets/GraphicTests/Scenes/9x_Other/9500_LightScripting/trasnparnetShadowCasting.mat]: name: transparent

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs (hunk 2 above)
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: yes, we should be able to drop it at this step. Was for 19.1 and nobody will go from 18.3 to 19.3
 :)

Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: I removed the version 1 upgrade code

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs (hunk 1 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/Light/HDAdditionalLightData.Migration.cs]: done

File: com.unity.render-pipelines.high-definition/Editor/Lighting/SerializedHDLight.cs [hunk 7]
```diff
@@ -68,128 +68,119 @@ public sealed class SerializedLightData
...
             public SerializedProperty contactShadows;
-            public SerializedProperty shadowTint;            
+            public SerializedProperty shadowTint;
             public SerializedProperty shadowUpdateMode;
 
             // Bias control
             public SerializedProperty constantBias;
 
             public SerializedProperty normalBias;
+
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/SerializedHDLight.cs]: fixed

File: com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.cs [hunk 8]
```diff
@@ -69,10 +67,17 @@ public override void OnInspectorGUI()
...
+            if (EditorGUI.EndChangeCheck())
+            {
+                m_SerializedHDLight.Apply();
 
-            m_SerializedHDLight.Apply();
+                foreach (var hdLightData in m_AdditionalLightDatas)
+                    hdLightData.UpdateAllLightValues();
+            }
 
+            // TODO: we don't need this anymore
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightEditor.cs]: fixed

**Comment by sebastienlagarde** [comment]
katana is green, other change inside the PR for the migratoin will be done in another PR

**Pull request merged**